File: location/src/test/java/io/appmetrica/analytics/location/impl/DataCacheTest.java
package io.appmetrica.analytics.location.impl;

import io.appmetrica.analytics.coreutils.internal.cache.DataCache;
import io.appmetrica.analytics.coreutils.internal.cache.LocationDataCacheUpdateScheduler;
import io.appmetrica.analytics.testutils.CommonTest;
import java.util.concurrent.TimeUnit;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public abstract class DataCacheTest<T> extends CommonTest {

    @Mock
    LocationDataCacheUpdateScheduler mDataCacheUpdateScheduler;

    private DataCache<T> mDataCache;
    private T mData;
    private T mAnotherData;

    private final long defaultCacheRefreshTime = TimeUnit.MINUTES.toMillis(1);
    private final long defaultCacheExpiryTime = TimeUnit.MINUTES.toMillis(2);

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);
        mDataCache = createDataCache(defaultCacheRefreshTime, defaultCacheExpiryTime);
        mData = createMockedData();
        mAnotherData = createAnotherData();
    }

    @Test
    public void testUpdateData() {
        mDataCache.updateData(mData);
        assertThat(mDataCache.getData()).isEqualTo(mData);
    }

    @Test
    public void testUpdateDataNullScheduler() {
        mDataCache.updateData(mData);
    }

    @Test
    public void testUpdateDataTriggersScheduler() {
        mDataCache.setUpdateScheduler(mDataCacheUpdateScheduler);
        mDataCache.updateData(mData);
        verify(mDataCacheUpdateScheduler).onStateUpdated();
    }

    @Test
    public void testGetData() {
        mDataCache.setUpdateScheduler(mDataCacheUpdateScheduler);
        mDataCache.updateData(mData);
        assertThat(mDataCache.getData()).isEqualTo(mData);
        verify(mDataCacheUpdateScheduler, never()).scheduleUpdateIfNeededNow();
    }

    @Test
    public void testGetDataIfShouldUpdate() throws InterruptedException {
        long refreshTime = 10L;
        mDataCache = createDataCache(refreshTime, refreshTime + 10);
        mDataCache.setUpdateScheduler(mDataCacheUpdateScheduler);
        mDataCache.updateData(mData);
        Thread.sleep(refreshTime + 2);
        assertThat(mDataCache.getData()).isEqualTo(mData);
        verify(mDataCacheUpdateScheduler).scheduleUpdateIfNeededNow();
    }

    @Test
    public void testGetDataIfShouldClear() throws InterruptedException {
        long refreshTime = 10L;
        long expiryTime = 12L;
        mDataCache = createDataCache(refreshTime, expiryTime);
        mDataCache.setUpdateScheduler(mDataCacheUpdateScheduler);
        mDataCache.updateData(mData);
        Thread.sleep(expiryTime + 1);
        assertThat(mDataCache.getData()).isNull();
        verify(mDataCacheUpdateScheduler).scheduleUpdateIfNeededNow();
    }

    @Test
    public void testGetDataIfCacheExpiredAndMissedUpdater() throws InterruptedException {
        mDataCache = createDataCache(10, 20);
        mDataCache.updateData(mData);
        Thread.sleep(10);
        mDataCache.getData();
    }

    @Test
    public void testGetDataIfCacheIsEmpty() {
        mDataCache.setUpdateScheduler(mDataCacheUpdateScheduler);
        mDataCache.getData();
        verify(mDataCacheUpdateScheduler).scheduleUpdateIfNeededNow();
    }

    @Test
    public void testShouldUpdateIfCacheIsEmpty() {
        assertThat(mDataCache.shouldUpdate()).isTrue();
    }

    @Test
    public void testShouldUpdateCacheIsOutdated() {
        long refreshTime = 10L;
        mDataCache = createDataCache(refreshTime, 20);
        mDataCache.updateData(mData);
        try {
            Thread.sleep(refreshTime + 1);
        } catch (InterruptedException e) {}
        assertThat(mDataCache.shouldUpdate()).isTrue();
    }

    @Test
    public void testShouldNotUpdate() {
        mDataCache.updateData(mData);
        assertThat(mDataCache.shouldUpdate()).isFalse();
    }

    @Test
    public void defaultRefreshTime() {
        assertThat(createDataCache().getCachedData().getRefreshTime()).isEqualTo(getDefaultCacheRefreshTime());
    }

    @Test
    public void defaultExpiryTime() {
        assertThat(createDataCache().getCachedData().getExpiryTime()).isEqualTo(getDefaultCacheExpiryTime());
    }

    @Test
    public void updateCacheControl() {
        long refreshTime = 123545;
        long expiryTime = 2312321L;
        mDataCache.updateCacheControl(refreshTime, expiryTime);
        assertThat(mDataCache.getCachedData().getRefreshTime()).isEqualTo(refreshTime);
        assertThat(mDataCache.getCachedData().getExpiryTime()).isEqualTo(expiryTime);
    }

    protected abstract DataCache<T> createDataCache(long refreshTime, long expiryTime);

    protected abstract DataCache<T> createDataCache();

    protected abstract long getDefaultCacheRefreshTime();

    protected abstract long getDefaultCacheExpiryTime();

    protected abstract T createMockedData();

    protected abstract T createAnotherData();
}

File: location/src/test/java/io/appmetrica/analytics/location/impl/DataCacheTest.kt
package io.appmetrica.analytics.location.impl

import io.appmetrica.analytics.coreutils.internal.cache.CachedDataProvider
import io.appmetrica.analytics.coreutils.internal.cache.DataCache
import io.appmetrica.analytics.coreutils.internal.cache.LocationDataCacheUpdateScheduler
import io.appmetrica.analytics.coreutils.internal.time.SystemTimeProvider
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.MockedConstructionRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner
import java.util.concurrent.TimeUnit

@RunWith(RobolectricTestRunner::class)
abstract class DataCacheTest<T> : CommonTest() {

    protected val dataCacheUpdateScheduler: LocationDataCacheUpdateScheduler = mock()

    protected lateinit var dataCache: DataCache<T>
    protected var data: T? = null
    protected var anotherData: T? = null

    private val defaultCacheRefreshTime = TimeUnit.MINUTES.toMillis(1)
    private val defaultCacheExpiryTime = TimeUnit.MINUTES.toMillis(2)

    @get:Rule
    val timeProviderRule = MockedConstructionRule(SystemTimeProvider::class.java)

    @get:Rule
    val cachedDataRule = MockedConstructionRule(CachedDataProvider.CachedData::class.java)

    @Before
    fun setUp() {
        dataCache = createDataCache(defaultCacheRefreshTime, defaultCacheExpiryTime)
        data = createMockedData()
        anotherData = createAnotherData()
    }

    protected fun cachedData(): CachedDataProvider.CachedData<T> {
        assertThat(cachedDataRule.constructionMock.constructed()).isNotEmpty
        val constructed = cachedDataRule.constructionMock.constructed()
        @Suppress("UNCHECKED_CAST")
        return constructed[constructed.size - 1] as CachedDataProvider.CachedData<T>
    }

    @Test
    fun updateData() {
        dataCache.updateData(data!!)

        verify(cachedData()).setData(data!!)
    }

    @Test
    fun updateDataNullScheduler() {
        dataCache.updateData(data!!)

        verify(cachedData()).setData(data!!)
    }

    @Test
    fun updateDataTriggersScheduler() {
        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.updateData(data!!)

        verify(cachedData()).setData(data!!)
        verify(dataCacheUpdateScheduler).onStateUpdated()
    }

    @Test
    fun getData() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(data!!)

        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.updateData(data!!)

        assertThat(dataCache.data).isEqualTo(data!!)
        verify(dataCacheUpdateScheduler, never()).scheduleUpdateIfNeededNow()
    }

    @Test
    fun getDataIfShouldUpdate() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(true)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(data!!)

        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.updateData(data!!)

        assertThat(dataCache.data).isEqualTo(data!!)
        verify(dataCacheUpdateScheduler).scheduleUpdateIfNeededNow()
    }

    @Test
    fun getDataIfShouldClear() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)
        whenever(cachedData().shouldClearData()).thenReturn(true)
        whenever(cachedData().data).thenReturn(null)

        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.updateData(data!!)

        assertThat(dataCache.data).isNull()
        verify(cachedData()).setData(null)
    }

    @Test
    fun getDataIfCacheExpiredAndMissedUpdater() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(true)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(data!!)

        dataCache.updateData(data!!)
        dataCache.data

        // No scheduler, so scheduleUpdateIfNeededNow should not be called
    }

    @Test
    fun getDataIfCacheIsEmpty() {
        whenever(cachedData().isEmpty).thenReturn(true)
        whenever(cachedData().shouldClearData()).thenReturn(false)
        whenever(cachedData().data).thenReturn(null)

        dataCache.setUpdateScheduler(dataCacheUpdateScheduler)
        dataCache.data

        verify(dataCacheUpdateScheduler).scheduleUpdateIfNeededNow()
    }

    @Test
    fun shouldUpdateIfCacheIsEmpty() {
        whenever(cachedData().isEmpty).thenReturn(true)

        assertThat(dataCache.shouldUpdate()).isTrue
    }

    @Test
    fun shouldUpdateCacheIsOutdated() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(true)

        dataCache.updateData(data!!)

        assertThat(dataCache.shouldUpdate()).isTrue
    }

    @Test
    fun shouldNotUpdate() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)

        dataCache.updateData(data!!)

        assertThat(dataCache.shouldUpdate()).isFalse
    }

    @Test
    fun updateCacheControl() {
        val refreshTime = 123545L
        val expiryTime = 2312321L

        dataCache.updateCacheControl(refreshTime, expiryTime)

        verify(cachedData()).setExpirationPolicy(refreshTime, expiryTime)
    }

    @Test
    fun shouldUpdateWhenTimeGoesBackwards() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(true)

        dataCache.updateData(data!!)

        assertThat(dataCache.shouldUpdate()).isTrue
    }

    @Test
    fun shouldClearWhenTimeGoesBackwards() {
        whenever(cachedData().isEmpty).thenReturn(false)
        whenever(cachedData().shouldUpdateData()).thenReturn(false)
        whenever(cachedData().shouldClearData()).thenReturn(true)
        whenever(cachedData().data).thenReturn(null)

        dataCache.updateData(data!!)

        assertThat(dataCache.data).isNull()
        verify(cachedData()).setData(null)
    }

    protected abstract fun createDataCache(refreshTime: Long, expiryTime: Long): DataCache<T>

    protected abstract fun createMockedData(): T

    protected abstract fun createAnotherData(): T
}
